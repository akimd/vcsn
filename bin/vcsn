#! /bin/sh

# This simple program gives an idea of what TAK-Kit v2 should look like.

prog="vcsn-$1"
shift
exec $prog "$@"
