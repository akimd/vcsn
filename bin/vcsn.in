#! /bin/sh

set -e

fatal ()
{
  local i
  for i
  do
    echo >&2 "$i"
  done
  exit 1
}

config ()
{
  # Call `vcsn-tools`, not `vcsn`, as `vcsn` is this very script...
  exec vcsn-tools configuration "configuration.$1"
}

# _Append_ VCSN_PYTHONDIR, not prepend, so that the user can still
# override this path via $PYTHONPATH.  Note in particular that the
# builddir-tests need to override $PYTHONPATH to use non-installed
# modules, which would them be overridden here and installed files
# would be used instead.
: ${VCSN_PYTHONDIR='@PYEXECDIR@'}
export VCSN_PYTHONDIR
PYTHONPATH=$PYTHONPATH:$VCSN_PYTHONDIR
export PYTHONPATH

# Also import our changes to PYTHONPATH.
MYPYPATH=$MYPYPATH:$PYTHONPATH:$VCSN_PYTHONDIR
export MYPYPATH

# Used by commands such as vcsn-notebook.  Avoids the need for a
# vcsn-notebook.in.
: ${VCSN_DATADIR='@DATADIR@'}
export VCSN_DATADIR

# Where our main executables are.
: ${VCSN_BINDIR='@BINDIR@'}
export VCSN_BINDIR

# Where our auxiliary executables are.
: ${VCSN_LIBEXECDIR='@LIBEXECDIR@'}
export VCSN_LIBEXECDIR

# If we are not installed, set the dynamic library path.
if test x"$VCSN_DYLD_LIBRARY_PATH" != x; then
  DYLD_LIBRARY_PATH=$VCSN_DYLD_LIBRARY_PATH
  export DYLD_LIBRARY_PATH
fi

# If there are environment variables that we wish to set, do it now.
# For instance VCSN_ENV=LD_PRELOAD=libclang_rt.asan-x86_64.so`.  This
# is especially useful on macOS with SIP enabled.
if test x"$VCSN_ENV" != x; then
  save_IFS=$IFS
  IFS=';'
  for v in $VCSN_ENV
  do
    eval "export $v"
  done
  IFS=$save_IFS
fi

# Find our auxiliary executables.  Also find our public tools: think
# for instance that we are not in the path, but the user called `vcsn`
# with a qualified path: we must be able to call `vcsn compile`
# etc. without requiring the user to put us in her path.
: ${VCSN_PATH=$VCSN_BINDIR:$VCSN_LIBEXECDIR:$PATH}
PATH=$VCSN_PATH
export PATH

export PYTHON=$(config python)
export IPYTHON=$(config ipython)

# The Tools commands.
commands ()
{
  exec vcsn-tools --commands
}

# The Tools commands as a grep pattern.
commands_pattern ()
{
  commands |
    tr -d '\n' |
    sed -e 's/^ */^(/;s/  */|/g;s/$/)$/'
}

help ()
{
  cat <<EOF
usage: vcsn COMMAND ARGS...

COMMAND can be:

- notebook: run Vcsn as an IPython Notebook.
- python, ipython: run Python/IPython.
- diagnose: check that Vcsn is properly installed.
- doc: open the documentation.
- gdb: run gdb with a type pretty-printer.

- compile: compile a C++ program using libvcsn.
- run: set up the environment to find Vcsn, and run the ARGS...
- ps: display Vcsn compilation jobs.
- demangle: improve readability of C++ compiler error messages.

- efstcompile, efstdecompile: compile/decompile EFST files

- version: display version tag.

- score, score-compare: generate/compare benchmarks.

- $(commands | sed -e '1s/^  //'):
  run the corresponding algorithm.

For more information, run:

  $ vcsn COMMAND -h

For instance:

  $ vcsn derived-term -C 'lal_char(abc), b' -e '(a+b+a)*' |
     vcsn determinize |
     vcsn minimize |
     vcsn to-expression
EOF
  exit 0
}

version ()
{
  authors=$(config authors)
  version=$(config version)
  url=$(config url)
cat <<EOF
vcsn (Vcsn $version)
Copyright (C) 2017 Vcsn Group.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

$(echo "Written by $authors." | fold -s -w 78)

For more information, see <$url>.
EOF
  exit 0
}

vcsn_doc_open ()
{
    jupyter=$(config jupyter)
    docdir=$(config docdir)
    nbdir=$docdir/notebooks
    if [[ ! -f "$nbdir/$1.ipynb" ]]; then
        fatal "vcsn doc: no documentation for $1"
    fi
    if "$jupyter" notebook --version >/dev/null; then
        exec "$jupyter" notebook "$nbdir/$1.ipynb"
    elif [[ -x /usr/bin/open ]]; then
        open "$nbdir/$1.html"
        exit 0
    else
        xdg-open "$nbdir/$1.html"
        exit 0
    fi
    fatal "vcsn doc: cannot open $nbdir/$1.html"
}

vcsn_doc ()
{
    case $1 in
        (-h|--help)
            cat <<EOF
Usage: vcsn doc [FUNCTION|PAGE]

FUNCTION is the name of the Python function, for instance

  \$ vcsn doc automaton.determinize

PAGE is a chapter of the documentation, for instance

  \$ vcsn doc Automata

By default open the '!Read-me-first' page.

If Jupyter notebook is installed, it will be used to display the
documentation, otherwise the plain HTML documentation is opened.
EOF
            exit 0;;

        (index) vcsn_doc_open 'Algorithms';;
        ('')    vcsn_doc_open '!Read-me-first';;
        (*)     vcsn_doc_open "$1";;
    esac
}


if test $# = 0; then
    fatal 'vcsn: missing argument' "Try 'vcsn --help' for more information."
fi

case $1 in
  ('--export') ;; # We are sourced to set up the environment.
  (run) shift 1; exec "$@";;
  (-h|--help) help;;
  (-v|--version) version;;
  (version) config version;;

  (compile|demangle|ps|score|score-compare)
      prog=$(command -v "vcsn-$1")
      shift
      exec "$PYTHON" "$prog" "$@";;

  (config)
      shift
      exec config "$@";;

  (check|diagnose|doctor)
      shift
      exec vcsn-diagnose "$@";;

  (doc)
      shift
      vcsn_doc "$@";;

  (efstcompile|efstdecompile)
      exec "$@";;

  (python) shift; exec "$PYTHON" "$@";;
  (ipython) shift; exec "$IPYTHON" "$@";;
  (python*|ipython*|jupyter*) exec "$@";;

  (*.ipynb) exec vcsn-notebook "$@";;
  (*.py)    exec "$PYTHON" "$@";;

  (*)
      if echo "$1" | grep -E "$(commands_pattern)" >/dev/null; then
          exec vcsn-tools "$@"
      else
          prog=vcsn-$1
          shift
          exec "$prog" "$@"
      fi;;
esac

# Local Variables:
# mode: shell-script
# End:
