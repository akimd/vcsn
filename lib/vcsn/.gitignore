/location.hh
/parse-rat-exp.cc
/parse-rat-exp.hh
/parse-rat-exp.html
/parse-rat-exp.output
/parse-rat-exp.xml
/position.hh
/stack.hh
