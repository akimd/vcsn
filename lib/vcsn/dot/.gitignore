/parse.cc
/parse.dir
/parse.hh
/parse.html
/parse.output
/parse.xml
/stack.hh
