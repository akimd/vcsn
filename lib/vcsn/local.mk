# Vaucanson 2, a generic library for finite state machines.
# Copyright (C) 2012-2013 Vaucanson Group.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# The complete GNU General Public Licence Notice can be found as the
# `COPYING' file in the root directory.
#
# The Vaucanson Group consists of people listed in the `AUTHORS' file.

# A Bison wrapper for C++.
BISONXX = $(top_builddir)/build-aux/bin/bison++
BISONXX_IN = $(top_srcdir)/build-aux/bin/bison++.in

BISONXXFLAGS =                                  \
  $(if $(V:0=),--verbose)
AM_BISONFLAGS =                                 \
  -Wall --report=all # -Werror

pkglib_LTLIBRARIES = lib/libvcsn.la
lib_libvcsn_la_SOURCES =                        \
  lib/vcsn/algos/accessible.cc                  \
  lib/vcsn/algos/aut-to-exp.cc                  \
  lib/vcsn/algos/complete.cc                    \
  lib/vcsn/algos/copy.cc                        \
  lib/vcsn/algos/de_bruijn.cc                   \
  lib/vcsn/algos/determinize.cc                 \
  lib/vcsn/algos/dot.cc                         \
  lib/vcsn/algos/edit-automaton.cc              \
  lib/vcsn/algos/eval.cc                        \
  lib/vcsn/algos/is_complete.cc                 \
  lib/vcsn/algos/is-deterministic.cc            \
  lib/vcsn/algos/is-eps-acyclic.cc              \
  lib/vcsn/algos/fsm.cc                         \
  lib/vcsn/algos/ladybird.cc                    \
  lib/vcsn/algos/lift.cc                        \
  lib/vcsn/algos/make-context.cc                \
  lib/vcsn/algos/print.cc                       \
  lib/vcsn/algos/product.cc                     \
  lib/vcsn/algos/read.cc                        \
  lib/vcsn/algos/registry.hh                    \
  lib/vcsn/algos/standard-of.cc                 \
  lib/vcsn/algos/transpose.cc                   \
  lib/vcsn/algos/xml.cc                         \
  lib/vcsn/ctx/ctx.cc                           \
  lib/vcsn/misc/escape.cc

# ratexpset.hxx depends on rat/driver.hh which requires
# rat/location.hh.  The dot parser, loads ratexpset.hxx, so we _must_
# compile the rat parser first.
lib/vcsn/dot/parser.stamp: lib/vcsn/rat/parser.stamp

include lib/vcsn/dot/local.mk
lib_libvcsn_la_SOURCES +=                       \
  $(SOURCES_DOT_PARSE_YY)                       \
  lib/vcsn/dot/driver.hh lib/vcsn/dot/driver.cc \
  lib/vcsn/dot/scan.ll

include lib/vcsn/rat/local.mk
lib_libvcsn_la_SOURCES +=                       \
  $(SOURCES_RAT_PARSE_YY)                       \
  lib/vcsn/rat/driver.hh lib/vcsn/rat/driver.cc \
  lib/vcsn/rat/scan.ll
