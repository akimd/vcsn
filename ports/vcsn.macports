# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4

PortSystem          1.0
PortGroup           cxx11 1.0
PortGroup           active_variants 1.1

name                vcsn
version             2.7
categories          devel
platforms           darwin
maintainers         lrde.epita.fr:akim openmaintainer
license             GPL-3+

description         C++ generic automata/transducers and rational expression platform

long_description    Vcsn is a platform for weighted automata and rational expressions. \
                    It consists of an efficient C++ generic library, shell tools, Python \
                    bindings, and a graphical interactive environment on top of IPython.

homepage            http://vcsn.lrde.epita.fr
master_sites        http://www.lrde.epita.fr/dload/vcsn/${version}/

use_xz              yes

checksums           rmd160  597f0ab44ad09bca405db33e2d48d3ac31d2ac9b \
                    sha256  95623ced01a312f60bb754ced6e9d143884e8f3063774ce6dba114147fd59618 \
                    size    14835124

# python3.x is required - force dependencies to use it as well
set python_version  36
set python_branch   [string range ${python_version} 0 end-1].[string index ${python_version} end]

depends_build-append    port:doxygen \
                        port:flex \
                        port:libtool

depends_lib-append  port:boost \
                    path:bin/ccache:ccache \
                    port:gmp \
                    port:python${python_version} \
                    port:yaml-cpp

depends_run         path:bin/dot:graphviz \
                    port:py${python_version}-ipython

require_active_variants boost python${python_version}

configure.args      --disable-silent-rules
configure.optflags  -O3 -DNDEBUG
configure.python    ${prefix}/bin/python${python_branch}

test.run            yes
test.target         check
