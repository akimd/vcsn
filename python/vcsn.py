from vcsn_python import *

def dot_to_svg(dot):
    open("/tmp/a.gv", "w").write(dot)
    from subprocess import check_call
    check_call(['dot','-Tsvg','/tmp/a.gv','-o','/tmp/a.svg'])
    return open("/tmp/a.svg", "r").read()

## --------- ##
## context.  ##
## --------- ##

def make_ratexp(ctx, re):
    return ratexp(ctx, re)
context.ratexp = make_ratexp
context._repr_latex_ = lambda c: '$' + c.format('latex') + '$'

## ----------- ##
## automaton.  ##
## ----------- ##

automaton._repr_svg_ = lambda a: dot_to_svg(str(a))
automaton.__mul__ = lambda lhs, rhs: lhs.product(rhs).trim()

def automaton_fst(aut, cmd):
    open("/tmp/in.efsm", "w").write(aut.format("efsm"))
    from subprocess import check_call, check_output
    check_call(['efstcompile',   '/tmp/in.efsm', '/tmp/in.fst'])
    check_call([cmd,             '/tmp/in.fst',  '/tmp/out.fst'])
    res = check_output(['efstdecompile', '/tmp/out.fst'])
    return automaton(res, "efsm")

automaton.fstdeterminize = lambda aut: automaton_fst(aut, "fstdeterminize")
automaton.fstminimize = lambda aut: automaton_fst(aut, "fstminimize")
