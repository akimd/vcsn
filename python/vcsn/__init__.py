import sys, traceback, inspect

from vcsn_cxx import (automaton, context, expansion, expression, label,
                      polynomial, weight)
from vcsn.automaton  import automaton
from vcsn.context    import context
from vcsn.expansion  import expansion
from vcsn.expression import expression
from vcsn.label      import label
from vcsn.polynomial import polynomial
from vcsn.weight     import weight

# Load IPython specific support if we can.
try:
    # Load only if we are running IPython.
    __IPYTHON__
    from vcsn.ipython import *
    from vcsn.score import *
except (ImportError, NameError):
    pass

from vcsn_tools.config import config # pylint: disable=wrong-import-order
datadir = config['datadir']
version = config['version']

# Predefined contexts for casual users.
B = context('lal_char, b')
Z = context('lal_char, z')
Q = context('lal_char, q')
b = B

# Cause an error to get back algorithms list
try:
    B.expression(r'\e').automaton('')
except RuntimeError as e:
    expression.automaton.algos = e.args[0].split('expected: ')[1].split()
try:
    B.expression(r'\e', '')
except RuntimeError as e:
    expression.identities_list = e.args[0].split('expected: ')[1].split()

# Traceback handler
_traceback_verbose = False

def is_vcsn(tb):
    '''Does this traceback frame contain vcsn code?'''
    try:
        return 'vcsn' == inspect.getmodule(tb.tb_frame).__package__
    except AttributeError:
        return False

def vcsn_traceback_levels(tb):
    '''Returns the amount of frames deep into the traceback
    we have to go through before we reach vcsn.'''
    length = 0
    while tb and not is_vcsn(tb):
        tb = tb.tb_next
        length += 1
    return length

def vcsn_exception_handler(type, value, tb):
    '''Prints traceback according to vcsn._traceback_verbose.'''
    if _traceback_verbose:
        traceback.print_exception(type, value, tb)
    else:
        length = vcsn_traceback_levels(tb)
        traceback.print_exception(type, value, tb, length)

sys.excepthook = vcsn_exception_handler
