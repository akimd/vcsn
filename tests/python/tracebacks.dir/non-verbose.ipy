import vcsn
def test():
    def test1():
        def test2():
            '{:f}'.format(vcsn.Q.expression('abc').automaton())
        test2()
    test1()
test()
