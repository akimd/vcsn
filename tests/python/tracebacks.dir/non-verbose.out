Traceback (most recent call last):
  File "tracebacks.dir/non-verbose.py", line 8, in <module>
    test()
  File "tracebacks.dir/non-verbose.py", line 7, in test
    test1()
  File "tracebacks.dir/non-verbose.py", line 6, in test1
    test2()
  File "tracebacks.dir/non-verbose.py", line 5, in test2
    '{:f}'.format(vcsn.Q.expression('abc').automaton())
RuntimeError: print: FAdo requires letter or nullable labels, and Boolean weights
